docs/design.html

```diff
@@ -136,8 +136,10 @@ <h4 class="anchor-heading"><a id="design_compression" class="anchor-link"></a><a
     the user can always compress its messages one at a time without any support needed from Kafka, but this can lead to very poor compression ratios as much of the redundancy is due to repetition between messages of
     the same type (e.g. field names in JSON or user agents in web logs or common string values). Efficient compression requires compressing multiple messages together rather than compressing each message individually.
     <p>
-    Kafka supports this with an efficient batching format. A batch of messages can be clumped together compressed and sent to the server in this form. This batch of messages will be written in compressed form and will
-    remain compressed in the log and will only be decompressed by the consumer.
+    Kafka supports this with an efficient batching format. A batch of messages can be grouped together, compressed, and sent to the server in this form. The broker decompresses the batch in order to validate it. For
+    example, it validates that the number of records in the batch is same as what batch header states. The broker may also potentially modify the batch (e.g., if the topic is compacted, the broker will filter out 
```

Comment:
I just realised another thing.

"if the topic is compacted, the broker will filter out records eligible for compaction prior to writing to disk"

Are you referring to the fact that records written to a compacted topic need to necessarily have a non-null key else they will be rejected? If yes, then perhaps, we need to phrase it differently.

Let me get back to you with a suggestion here in a couple of hours.

Reply:
thank you! sounds good

Reply:
Thanks for checking in on this-- in light of the latest comment,  I removed the sentence "The broker may also potentially modify the batch (e.g., if the topic is compacted, the broker will filter out 
    records eligible for compaction prior to writing to disk)."

Reply:
@divijvaidya @mjsax won't the broker decompress async to perform compaction filtering? ([here](https://github.com/apache/kafka/blob/a7e865c0a756504cc7ae6f4eb0772cadd3333c53/core/src/main/scala/kafka/log/LogCleaner.scala#L801))

Reply:
Well, that's on the log-cleaner thread, so when actual compaction happens.

This part of the docs is focusing on the write path though.